### PR TITLE
Add keys, values, entires functions for typedArray

### PR DESCRIPTION
**Pull request by szilagyiadam**
Extended the typedArray functionality based on ES2015

Co-authored-by: Tibor Dusnoki tdusnoki@inf.u-szeged.hu
JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

File: tests/jerry/es2015/typedarray-prototype-keys.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+try {
+  var array = [Uint8Array, Uint16Array, Uint32Array, 
+               Float32Array, Float64Array, 
+               Int8Array, Int16Array, Int32Array];
+
+  array.forEach (function(e){
+    e.prototype.keys.call (undefined);
```
Comment (galpeter) [inline tests/jerry/es2015/typedarray-prototype-keys.js]: This test is not correct. If the first (`Uint8Array`) `keys` call is performed and the expected output is a TypeError which is catched below then the second (and further) cases are not tested. The iteration of types should be outside of the try catch block and only the `..keys.call(..)` should be wrapped with try-catch. This is same for all test files.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 2]
```diff
@@ -312,6 +313,95 @@ ecma_builtin_typedarray_prototype_for_each (ecma_value_t this_arg, /**< this arg
...
+  {
+    return this_obj;
+  }
+
+  ecma_object_t *obj_p = ecma_get_object_from_value (this_obj);
+
+  ecma_object_t *prototype_obj_p = ecma_builtin_get (ECMA_BUILTIN_ID_ARRAY_ITERATOR_PROTOTYPE);
+
+  ecma_value_t ret_value = ecma_op_create_iterator_object (this_obj,
+                                                            prototype_obj_p,
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: Incorrect indentation.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 3]
```diff
@@ -312,6 +313,95 @@ ecma_builtin_typedarray_prototype_for_each (ecma_value_t this_arg, /**< this arg
...
+ *      Returned value must be freed with ecma_free_value.
+ *
+ * @return iterator result object, if success
+ *         error - otherwise
+ */
+static ecma_value_t
+ecma_builtin_typedarray_iterators_helper (ecma_value_t this_arg, /**< this argument */
+                                          uint8_t type) /**< any combination of ecma_iterator_type_t bits */
+{
+  ecma_value_t this_obj = ecma_op_to_object (this_arg);
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: There is a conversion here, shouldn't the `this_obj` be freed at the end??

**Comment by szilagyiadam** [comment]
Did the requested changes.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.inc.h [hunk 4]
```diff
@@ -71,6 +71,14 @@ ROUTINE (LIT_MAGIC_STRING_SORT, ecma_builtin_typedarray_prototype_sort, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_FIND, ecma_builtin_typedarray_prototype_find, 2, 1)
 ROUTINE (LIT_MAGIC_STRING_FIND_INDEX, ecma_builtin_typedarray_prototype_find_index, 2, 1)
 
+#if ENABLED (JERRY_ES2015_BUILTIN_ITERATOR)
+
+ROUTINE (LIT_MAGIC_STRING_KEYS, ecma_builtin_typedarray_prototype_keys, 0, 0)
+ROUTINE (LIT_MAGIC_STRING_VALUES, ecma_builtin_typedarray_prototype_values, 0, 0)
+ROUTINE (LIT_MAGIC_STRING_ENTRIES, ecma_builtin_typedarray_prototype_entries, 0, 0)
+
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.inc.h]: `LIT_GLOBAL_SYMBOL_ITERATOR` should be added as well.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c (hunk 3 above)
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: The `ecma_deref_object (obj_p);` call stands for freeing `this_obj`.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 5]
```diff
@@ -312,6 +313,96 @@ ecma_builtin_typedarray_prototype_for_each (ecma_value_t this_arg, /**< this arg
...
+
+  if (ECMA_IS_VALUE_ERROR (this_obj))
+  {
+    return this_obj;
+  }
+
+  ecma_object_t *obj_p = ecma_get_object_from_value (this_obj);
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: The `obj_p` also need to be checked whether its a `TypedArray` instance or not.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 6]
```diff
@@ -312,6 +313,101 @@ ecma_builtin_typedarray_prototype_for_each (ecma_value_t this_arg, /**< this arg
...
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_typedarray_prototype_keys (ecma_value_t this_arg) /**< this argument */
+{
+  return ecma_builtin_typedarray_iterators_helper (this_arg, ECMA_ITERATOR_KEYS);
+} /* ecma_builtin_typedarray_prototype_keys */
+
+/**
+ * The %TypedArray%.prototype object's 'values' routine
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: Please also mention the `@@iterator` routine in the description as well

File: tests/jerry/es2015/typedarray-prototype-entires.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+// limitations under the License.
+
+var array = [Uint8Array, Uint16Array, Uint32Array, 
+             Float32Array, Float64Array, 
+             Int8Array, Int16Array, Int32Array];
+
+array.forEach (function(e){
+  try {
+    e.prototype.entries.call (undefined);
+     assert (false);
```
Comment (rerobika) [inline tests/jerry/es2015/typedarray-prototype-entires.js]: This line is misaligned (same in the others tests as well)

File: tests/jerry/es2015/typedarray-prototype-entires.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    assert (e.message === "foo");
+  }
+}
+
+current_item = iterator.next ();
+assert (current_item.value === undefined);
+assert (current_item.done === true);
+
+/* Test empty array */
+array = [];
```
Comment (rerobika) [inline tests/jerry/es2015/typedarray-prototype-entires.js]: This should be an empty TypedArray.

File: tests/jerry/es2015/typedarray-prototype-entires.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+iterator = array.entries ();
+current_item = iterator.next ();
+
+assert (current_item.value === undefined);
+assert (current_item.done === true);
+
+/* Test delete elements after the iterator has been constructed */
+
+array = [0, 1, 2, 3, 4, 5];
```
Comment (rerobika) [inline tests/jerry/es2015/typedarray-prototype-entires.js]: Ditto.

File: tests/jerry/es2015/typedarray-prototype-values.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+});
+
+var buffer = new ArrayBuffer(8);
+var array = new Uint8Array(buffer);
+array.set([1, 2, 3, 4, 5, 6, 7, 8]);
+
+var iterator = array.values ();
+
+/* The initial value of the @@iterator property is the same function object
+ as the initial value of the Array.prototype.values property. */
```
Comment (rerobika) [inline tests/jerry/es2015/typedarray-prototype-values.js]: `Array.prototype.values` -> `%TypedArray%.prototype.values`

File: tests/jerry/es2015/typedarray-prototype-keys.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    assert (false);
+  }
+  catch (e) {
+    assert (e instanceof TypeError)
+  }
+});
+
+var buffer = new ArrayBuffer(8);
+var array = new Uint8Array(buffer);
```
Comment (galpeter) [inline tests/jerry/es2015/typedarray-prototype-keys.js]: What about the other TypedArrays? Shouldn't we test them also?

**Comment by szilagyiadam** [comment]
Updated the pl.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 12]
```diff
@@ -312,6 +313,104 @@ ecma_builtin_typedarray_prototype_for_each (ecma_value_t this_arg, /**< this arg
...
+ *      Returned value must be freed with ecma_free_value.
+ *
+ * @return iterator result object, if success
+ *         error - otherwise
+ */
+static ecma_value_t
+ecma_builtin_typedarray_iterators_helper (ecma_value_t this_arg, /**< this argument */
+                                          uint8_t type) /**< any combination of ecma_iterator_type_t bits */
+{
+
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: Unnecessary new line.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c (hunk 3 above)
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: This arg must be an object so the conversion to object is unnecessary.

File: jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c [hunk 13]
```diff
@@ -312,6 +313,104 @@ ecma_builtin_typedarray_prototype_for_each (ecma_value_t this_arg, /**< this arg
...
+                                                           prototype_obj_p,
+                                                           ECMA_PSEUDO_ARRAY_ITERATOR,
+                                                           type);
+  ecma_deref_object (obj_p);
+
+  return ret_value;
+} /* ecma_builtin_typedarray_iterators_helper */
+
+/**
+ * The %TypedArray%.prototype object's 'keys' routine using @@iterator
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c]: Could you explain to me the meaning of `usi ng @@iterator`?

**Pull request merged**